langkit/support/langkit_support-diagnostics-output.adb

```diff
@@ -0,0 +1,204 @@
+------------------------------------------------------------------------------
+--                                                                          --
+--                                 Langkit                                  --
+--                                                                          --
+--                       Copyright (C) 2020, AdaCore                        --
+--                                                                          --
+-- Langkit is free software; you can redistribute it and/or modify it under --
+-- terms of the  GNU General Public License  as published by the Free Soft- --
+-- ware Foundation;  either version 3,  or (at your option)  any later ver- --
+-- sion.   This software  is distributed in the hope that it will be useful --
+-- but WITHOUT ANY WARRANTY;  without even the implied warranty of MERCHAN- --
+-- TABILITY  or  FITNESS  FOR A PARTICULAR PURPOSE.                         --
+--                                                                          --
+-- As a special  exception  under  Section 7  of  GPL  version 3,  you are  --
+-- granted additional  permissions described in the  GCC  Runtime  Library  --
+-- Exception, version 3.1, as published by the Free Software Foundation.    --
+--                                                                          --
+-- You should have received a copy of the GNU General Public License and a  --
+-- copy of the GCC Runtime Library Exception along with this program;  see  --
+-- the files COPYING3 and COPYING.RUNTIME respectively.  If not, see        --
+-- <http://www.gnu.org/licenses/>.                                          --
+------------------------------------------------------------------------------
+
+with Ada.Strings.Wide_Wide_Fixed; use Ada.Strings.Wide_Wide_Fixed;
+with Ada.Text_IO;                 use Ada.Text_IO;
+with Ada.Wide_Wide_Text_IO;       use Ada.Wide_Wide_Text_IO;
+
+with GNATCOLL.Terminal;           use GNATCOLL.Terminal;
+
+with Langkit_Support.Images;      use Langkit_Support.Images;
+
+package body Langkit_Support.Diagnostics.Output is
+
+   Term_Info   : Terminal_Info;
+   Colors_Init : Boolean := False;
+
+   use Ada.Containers;
+
+   procedure Print_Source_Listing
+     (Sloc_Range  : Source_Location_Range;
+      Buffer      : in out Text_Buffer;
+      Lines_After : Natural := 0);
+   --  Print a source listing
+
+   procedure Reset_Colors;
+   --  Reset the state of colors for ``Term_Info``
+
+   ------------------
+   -- Reset_Colors --
+   ------------------
+
+   procedure Reset_Colors is
+   begin
+      Set_Color (Term_Info, Foreground => Reset);
+      Set_Style (Term_Info, Reset_All);
+   end Reset_Colors;
+
+   --------------
+   -- Get_Line --
+   --------------
+
+   function Get_Line
+     (Self        : in out Text_Buffer;
+      Line_Number : Positive) return Text_Type
+   is
+      T      : Text_Type renames Self.Text_Ptr.all;
+      Offset : Natural := 0;
```

Comment:
I realize only now that this should be initialized to `T'First`. And thus probably that the variable should be renamed, as it’s more an index than an offset.